File: conj_grad.py
import copy
import numpy as np
from numpy import linalg as la
from sklearn.covariance import EmpiricalCovariance as Ecov
import time
from matplotlib import pyplot as plt

def conj_grad(A,b):
    """perform conjugate gradient descent on Ax=b until 
    error tolerance is reached"""
    [m,n] = np.shape(A)
    #initialize state vectors
    x = np.zeros([m,1])
    y = copy.copy(x)
    #initialize residual
    r = b - np.dot(A,x)
    #initialize transpose of residual
    rsold = np.dot(np.transpose(r),r)
    #initialize search direction
    p = r
    #calculate A*p vector
    A_p = np.dot(A,p)
    #normalizing constant
    d = np.dot(np.transpose(p),np.dot(A,p))

    #########################################
    # STORE P and d for covariance analysis #
    #########################################

    P_matrix = np.array([])
    d_vector = np.array([])

    #iterate m times (or until convergence)
    for i in range(m):
        print("iteration {}/{}".format(i,m))
        gamma = (rsold)/d
        x = x + np.dot(p,gamma)
        z = np.random.randn()
        y = y + (z/np.sqrt(d)) * p
        #store old r to calculate new beta
        r = r - gamma*A_p
        rsnew = np.dot(np.transpose(r),r)
        #is the magnitude of the residual < 1e-16?
        if np.sqrt(rsnew) < 1e-8:
            print("converged at iteration {}/{}".format(i,m))
        #calculate new beta
        beta = rsnew/rsold
        #update rsold constant
        rsold = rsnew
        #store P matrix and d vector entry
        if i==0:
            P_matrix = p
        else:
            P_matrix = np.append(P_matrix,p,axis=1)
        print("p is {}, p_matrix is {}".format(p,P_matrix))
        d_vector = np.append(d_vector,d)
        print("d_vector is {}".format(d_vector))
        #calculate new search direction
        p = r + beta*p
        #calculate A*p vector
        A_p = np.dot(A,p)
        #calculate new normalization constant
        d = np.dot(np.transpose(p),A_p)
    
    #########################################
    # make analytic covariance matrix       #
    #########################################
 
    d_matrix = np.diag(np.transpose(d_vector))
    print("shape of P is {}, shape of D is {}".format(np.shape(P_matrix),np.shape(d_matrix)))
    CG_cov = np.matmul(P_matrix, np.matmul(la.inv(d_matrix),np.transpose(P_matrix)))
    cov = la.inv(A)
    rel_err = la.norm(cov - CG_cov)/la.norm(cov)
    print("relative error in analytical covariance matrix approximation is {}".format(rel_err))
    return x,y,CG_cov

def cg_plot(A,b):
    
    #########################################
    # solver solution plot                  #
    #########################################
    cov = la.inv(A) 
    x_a = np.matmul(cov,b)
    x_cg,y,CG_cov = conj_grad(A,b)
    x_a = np.sort(np.ravel(x_a))
    x_cg = np.sort(np.ravel(x_cg))
    rel_err = la.norm(x_a - x_cg)/la.norm(x_a)
    plt.plot(x_a,label="analytic solution")
    plt.plot(x_cg,marker='1',linestyle='none',label='CG solution')
    plt.legend()
    plt.title("Analytic solution and cg solution for Ax=b, rel err {}".format(rel_err))
    plt.show()
    
    #########################################
    #    COVARIANCE EIGENVALUE PLOT         #
    #########################################
    rel_err = la.norm(cov - CG_cov)/la.norm(cov)
    eigs, eigvecs = la.eigh(cov)
    plt.semilogy(eigs,label="actual eigenvalues")
    eigs_CG, eigvecs_CG = la.eigh(CG_cov)
    #masking invalid CG eigenvalues
    eigs_CG = np.ma.masked_where(eigs_CG<1e-5,eigs_CG)
    plt.semilogy(eigs_CG,marker='1',linestyle='none',label='CG empirical eigenvalues')
    plt.legend()
    plt.title("CG sampler cov evals vs actual evals. Relative error in cov matrix approximation {}".format(rel_err))
    plt.show()

if __name__=="__main__":
    A = np.loadtxt("test_A.txt",delimiter=',')
    [dims,dims] = np.shape(A)
    b = np.random.randn(dims,1)
    cg_plot(A,b)
    A = np.loadtxt("2d.txt",delimiter=',')
    [dims,dims] = np.shape(A)
    b = np.random.randn(dims,1)
    cg_plot(A,b)
    A = np.loadtxt("3d.txt",delimiter=',')
    [dims,dims] = np.shape(A)
    b = np.random.randn(dims,1)
    cg_plot(A,b)

File: conj_grad.m
function [x,y,c,count,lost_conj] = conj_grad(A,b,iters)
    %make initial guess
    [m,n] = size(A);
    x = zeros(m,1);
    %initialize sample as well
    y = zeros(m,1);
    %initialize c_k
    c = zeros(m,1);
    %initial residual
    r = b - A*x;
    %initial search direction
    p = r;
    %normalizing constant
    d = p'*A*p;
    %cov = zeros(m,m);
    %min_rel_err = 1.0;
    %min_c = zeros(m,1);
    %min_iter = 1;
    p_vec = [];
    for count=1:iters
        gamma = (r'*r)/d;
        %calculating state vector
        x = x + gamma*p;
        %sample z ~ N(0,1)
        z = randn();
        %update sample
        y = y + (z/sqrt(d))*p;
        %update c_k, storing old c_k
        %c = c + (z/sqrt(d))*(A*p);
        %store old r to calculate new beta
        r_old = r;
        r = r_old - gamma*A*p;
        %standard beta
        %beta = -(r'*r)/(r_old'*r_old);
        %polak-ribiere method
        beta = -(r'*(r - r_old))/(r_old'*r_old);
        p_old = p;
        p = r - beta*p;
        d = p'*A*p;
        %add new p vector to set 
        p_vec = [p_vec p];
        %calculate A-conjugacy of p vectors so far
        conj_vec = [];
        for i=1:(count-1)
            p_i = p_vec(:,i);
            conj = dot(p_i,A*p);
            conj_vec(i) = conj;
        end
        mean_conj = mean(conj_vec);
        if mean_conj > 1e-2
            lost_conj = count+1;
            sprintf('lost conjugacy at iteration %d',lost_conj)
            break;
        end
        sprintf('mean A conjugacy at iter %d is: %d',count+1,mean(conj_vec));
        %calculate covariance at this point
        %cov_approx_part = (A)*(1/d)*(p*p')*(A);
        %cov = cov + cov_approx_part;
        %A;
        %rel_err = norm(cov - A)/norm(A);
        %sprintf('rel error %d at iter %d, with (1/d)=%d',rel_err,count,1/d);
        %if rel_err < min_rel_err
        %    min_c = c;
        %    min_iter = count;
        %    min_rel_err = rel_err;
        %elseif rel_err > (1.5*min_rel_err)
        %    sprintf('going uphill at iter %d, min err %d current error %d ',count,min_rel_err,rel_err)
        %    c = min_c;
        %    break;
        %end
        %check for convergence
        %if d<=1e-8
        %    sprintf('d diverged at iteration %d with d=%d',count,abs(d))
        %     %c = A*y;
        %    break;
        if norm(r)<1e-4
            sprintf('r converged to %d at iteration %d',norm(r),count)
            c = A*y;
            break;
        end
    end
    c = A*y;
    sprintf('exited loop at iter count');
end